src/timeout.py

```diff
@@ -20,96 +20,164 @@
-            return node
-        failed_child_nodes = self._db.count_nodes({
-            'parent': node['_id'],
-            'result': 'fail'
-        })
-        node['result'] = 'fail' if failed_child_nodes else 'pass'
-        return node
-
-    def _update_available_node(self, node, current_time):
-        """Set node state to 'done' if holdoff is expired and all child nodes
-        are completed.
-        Set node state to `closing` if holdoff is expired and at least one of
-        the child nodes is incomplete"""
-        holdoff_time = datetime.fromisoformat(node['holdoff'])
-        if current_time > holdoff_time:
-            self._logger.log_message(logging.INFO,
-                                     f"Holdoff reached for {node['_id']}")
-            total_child_nodes = self._db.count_nodes({
-                'parent': node['_id'],
+    def _get_pending_nodes(self, filters=None):
+        nodes = {}
+        node_filters = filters.copy() if filters else {}
+        for state in ['running', 'available', 'closing']:
+            node_filters['state'] = state
+            for node in self._db.get_nodes(node_filters):
+                nodes[node['_id']] = node
+        return nodes
+
+    def _count_running_child_nodes(self, parent_id):
+        nodes_count = 0
+        for state in ['running', 'available', 'closing']:
+            nodes_count += self._db.count_nodes({
+                'parent': parent_id, 'state': state
             })
+        return nodes_count
+
+    def _get_child_nodes_recursive(self, node, state_filter=None):
+        recursive = {}
+        child_nodes = self._get_pending_nodes({'parent': node['_id']})
+        for child_id, child in child_nodes.items():
+            if state_filter is None or child['state'] == state_filter:
+                recursive.update(self._get_child_nodes_recursive(
+                    child, state_filter
+                ))
+        return recursive
+
+    def _submit_lapsed_nodes(self, lapsed_nodes, state, log=None):
+        for node_id, node in lapsed_nodes.items():
+            node_update = node.copy()
+            node_update['state'] = state
+            if log:
+                self.log.debug(f"{node_id} {log}")
+            self._db.submit({'node': node_update})
+
+
+class Timeout(TimeoutService):
+
+    def __init__(self, configs, args):
+        super().__init__(configs, args, 'timeout')
+
+    def _check_pending_nodes(self, pending_nodes):
+        now = datetime.utcnow()
+        sleep = timedelta(seconds=self._poll_period)
+        timeout_nodes = {}
+        for node_id, node in pending_nodes.items():
+            timeout = datetime.fromisoformat(node['timeout'])
+            if now > timeout:
+                timeout_nodes[node_id] = node
+                timeout_nodes.update(self._get_child_nodes_recursive(node))
+            else:
+                self.log.debug(f"{node_id} timeout left: {timeout - now}")
+                sleep = min(sleep, timeout - now)
```

Comment:
So, here polling period depends on the node's timeout value as well.
This way timed-out nodes will be updated more frequently than they would be in case of the actual polling period.
Won't it increase iterations?
What are the pros of doing this?

Reply:
It just means, sleep for 60 seconds unless we know a node is going to timeout earlier than that.  It allows the sleep period to be longer as we can "wake up" in time for the next timeout.  This works really well with our current use-case as we have a handful of nodes to monitor and they normally don't time out.  It does mean that when a node times out or completes less than 60s before its timeout time, the query to get the pending nodes gets sent earlier as a side effect.  To scale better, we should look into subscribing to events about nodes entering the Available state and make use of the timeout timestamp when querying for pending nodes to retrieve only the ones that matter (I think I explained that in a comment somewhere).

Reply:
Yes, after the PR https://github.com/kernelci/kernelci-api/pull/197 gets merged,
we can use filter {'timeout__lt': datetime.isoformat(datetime.utcnow())}

Reply:
Yes, I'll create some GitHub issues with some ideas for optimising the timeout implementation.